### PR TITLE
i#2364: dr_insert_mbr_instrumentation on iret

### PR DESCRIPTION
**Pull request by Simorfo**
Sets right size to operand in dr_insert_mbr_instrumentation.

Fixes 2364

### CODE REVIEW

**Comment by Simorfo** [comment]
I did not know what to do for the test.
I chose to create a new file, inspired by client-retaddr

File: core/lib/instrument.c [hunk 1]
```diff
@@ -5920,7 +5920,11 @@ dr_insert_mbr_instrumentation(void *drcontext, instrlist_t *ilist, instr_t *inst
         /* the retaddr operand is always the final source for all OP_ret* instrs */
         opnd_t retaddr = instr_get_src(instr, instr_num_srcs(instr) - 1);
         opnd_size_t sz = opnd_get_size(retaddr);
-        /* even for far ret and iret, retaddr is at TOS */
+        /* even for far ret and iret, retaddr is at TOS
```
Comment (derekbruening) [inline core/lib/instrument.c]: style nit: capitalize sentence

File: suite/tests/client-interface/retinstru.dll.c [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+#include "dr_api.h"
+
+uint num_ret = 0;
+
+static void mbr_instru(app_pc instr_addr, app_pc target_addr) {
+    num_ret++;
```
Comment (derekbruening) [inline suite/tests/client-interface/retinstru.dll.c]: It would be better to have some kind of check that the right values are passed as the parameters here, if possible.

File: suite/tests/client-interface/retinstru.c [hunk 3]
```diff
@@ -0,0 +1,172 @@
...
+        pop      REG_XBX
+        ret
+     skip_far:
+#ifdef X64
+        RAW(48) /* yeah the default is ridiculously 4 bytes */
+#endif
+        RAW(cb) /* far ret */
+        END_FUNC(FUNCNAME)
+
+END_FILE
```
Comment (derekbruening) [inline suite/tests/client-interface/retinstru.c]: This looks identical to retaddr.c.  Given that the test in the .dll.c is only counting and not checking the params, I would say let's not add a new identical app and just add the mbr code to retaddr.dll.c (where it will also function as a test of combining the mbr with the zero-retaddr instru).  (Or, add it to the count-ctis test which is what is already doing this kind of test for mbr instru, along with an iret in count-ctis.c.)

File: core/lib/instrument.c (hunk 1 above)
Comment (Simorfo) [inline core/lib/instrument.c]: Ok, I did not dare to change what was already there.

File: suite/tests/client-interface/retinstru.c (hunk 3 above)
Comment (Simorfo) [inline suite/tests/client-interface/retinstru.c]: Ok, I will then add it to retaddr which is the one where I found the bug first

File: suite/tests/client-interface/retinstru.dll.c (hunk 2 above)
Comment (Simorfo) [inline suite/tests/client-interface/retinstru.dll.c]: I agree but I did not see how to do it since there are more mbrs than the ones we want to test for.

File: core/lib/instrument.c (hunk 1 above)
Comment (derekbruening) [inline core/lib/instrument.c]: You were already changing that line.

File: suite/tests/client-interface/retaddr.dll.c [hunk 4]
```diff
@@ -32,20 +32,42 @@
 
 #include "dr_api.h"
 
+uint num_ret = 0;
+
+static void mbr_instru(app_pc instr_addr, app_pc target_addr) {
```
Comment (derekbruening) [inline suite/tests/client-interface/retaddr.dll.c]: style violation: void\n

File: suite/tests/client-interface/retaddr.dll.c [hunk 5]
```diff
@@ -32,20 +32,42 @@
...
             dr_clobber_retaddr_after_read(drcontext, bb, instr, 0);
+            /* i#2364 : mbr should be tested against different kinds of return. */
+            dr_insert_mbr_instrumentation(drcontext, bb, instr,
+                                          (app_pc)mbr_instru, SPILL_SLOT_1);
+        }
     }
     return DR_EMIT_DEFAULT;
 }
 
+void exit_event(void)
```
Comment (derekbruening) [inline suite/tests/client-interface/retaddr.dll.c]: style violation: void\n

Should also be static

File: suite/tests/client-interface/retaddr.dll.c [hunk 6]
```diff
@@ -32,20 +32,42 @@
...
+{
+    /* We assume returns are executed at least twice. */
+    if (num_ret > 1) {
+        dr_fprintf(STDERR, "instrumentation for return ok\n");
+    }
+    else {
+        dr_fprintf(STDERR, "FAIL no instrumentated returns\n");
```
Comment (derekbruening) [inline suite/tests/client-interface/retaddr.dll.c]: nit: "instrumented"

File: suite/tests/client-interface/retaddr.dll.c [hunk 7]
```diff
@@ -34,7 +34,8 @@
 
 uint num_ret = 0;
 
-static void mbr_instru(app_pc instr_addr, app_pc target_addr) {
+static void
+mbr_instru(app_pc instr_addr, app_pc target_addr) {
```
Comment (derekbruening) [inline suite/tests/client-interface/retaddr.dll.c]: style: { needs to be on own line

**Comment by derekbruening** [comment]
Thanks for fixing this!
